### ISSUE TITLE
POST /self-service/registration route not accepting CSRF token

### ISSUE THREAD

**Issue by brycesteinhoff** [ory/kratos#2003]
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/kratos/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/kratos/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Cloud](https://www.ory.sh/) project.
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe the bug

I'm attempting to call the POST /self-service/registration endpoint from a backend Go application using the Go SDK.

I am setting the CSRF token on the SubmitSelfServiceRegistrationFlowBody like so:

```
submitSelfServiceRegistrationFlowWithPasswordMethodBody := kratos.NewSubmitSelfServiceRegistrationFlowWithPasswordMethodBody(
	"password",
	password,
	traits,
)

// CSRF token set
submitSelfServiceRegistrationFlowWithPasswordMethodBody.SetCsrfToken(csrfToken)

submitSelfServiceRegistrationFlowBody := kratos.SubmitSelfServiceRegistrationFlowBody{
	SubmitSelfServiceRegistrationFlowWithPasswordMethodBody: submitSelfServiceRegistrationFlowWithPasswordMethodBody,
}
```

I can see that if I `json.Marshal` the `submitSelfServiceRegistrationFlowWithPasswordMethodBody` that the `csrf_token` field is on the object as expected.

However when I execute the submission flow, I get back a CSRF error:

```
successfulSelfServiceRegistration, httpResponse, err := client.kratosPublicClient.V0alpha2Api.
	SubmitSelfServiceRegistrationFlow(context.Background()).
	Flow(flowId).
	SubmitSelfServiceRegistrationFlowBody(submitSelfServiceRegistrationFlowBody).
	Execute()
```

```
{{"error":{"id":"security_csrf_violation","code":403,"status":"Forbidden","request":"a9d454e8-56b8-4545-bbf1-dbfe94e1231c","reason":"The request was rejected to protect you from Cross-Site-Request-Forgery (CSRF) which could cause account takeover, leaking personal information, and other serious security issues.","details":{"docs":"https://www.ory.sh/kratos/docs/debug/csrf","hint":"The anti-CSRF cookie was found but the CSRF token was not included in the HTTP request body (csrf_token) nor in the HTTP Header (X-CSRF-Token).","reject_reason":"The HTTP Cookie Header was set and a CSRF token was sent but they do not match. We recommend deleting all cookies for this domain and retrying the flow."},"message":"the request was rejected to protect you from Cross-Site-Request-Forgery"}}
```

The value I'm setting is the value of the CSRF token cookie received back from prior `InitializeSelfServiceRegistrationFlowForBrowsers` request.

Do I need to use the name of the cookie also somehow? Or only the value in this field?

### Reproducing the bug

See code above

### Version

v0.8.0-alpha.2

### On which operating system are you observing this issue?

Linux

### In which environment are you deploying?

Kubernetes with Helm

**Comment by brycesteinhoff** [comment]
I added some outbound request logging on my Go app container.

I see that the request to Kratos (POST /self-service/registration) includes the `csrf_token` in the body. There is no `Cookie` header with the request (presumably not needed or added by the Go client SDK, and not documented [in the API reference here](https://www.ory.sh/kratos/docs/reference/api/#operation/submitSelfServiceRegistrationFlow)).

```
2021-11-28T02:53:01.933710Z	critical	envoy lua	script log:  - :authority: kratos-public.identity.svc.cluster.local
2021-11-28T02:53:01.933727Z	critical	envoy lua	script log:  - :path: /self-service/registration/browser
2021-11-28T02:53:01.933731Z	critical	envoy lua	script log:  - :method: GET
2021-11-28T02:53:01.933733Z	critical	envoy lua	script log:  - :scheme: http
2021-11-28T02:53:01.933735Z	critical	envoy lua	script log:  - user-agent: OpenAPI-Generator/1.0.0/go
2021-11-28T02:53:01.933737Z	critical	envoy lua	script log:  - accept: application/json
2021-11-28T02:53:01.933747Z	critical	envoy lua	script log:  - accept-encoding: gzip
2021-11-28T02:53:01.933751Z	critical	envoy lua	script log:  - x-forwarded-proto: http
2021-11-28T02:53:01.933755Z	critical	envoy lua	script log:  - x-request-id: 0681f47b-aea0-9153-9132-e9ecc60e6ff3
2021-11-28T02:53:01.933757Z	critical	envoy lua	script log:  - x-envoy-decorator-operation: kratos-public.identity.svc.cluster.local:80/*
2021-11-28T02:53:01.942034Z	critical	envoy lua	script log:  - :authority: kratos-public.identity.svc.cluster.local
2021-11-28T02:53:01.942052Z	critical	envoy lua	script log:  - :path: /self-service/registration?flow=4bdbbfcb-9faf-46ab-9449-0a149f7aedca
2021-11-28T02:53:01.942055Z	critical	envoy lua	script log:  - :method: POST
2021-11-28T02:53:01.942057Z	critical	envoy lua	script log:  - :scheme: http
2021-11-28T02:53:01.942059Z	critical	envoy lua	script log:  - user-agent: OpenAPI-Generator/1.0.0/go
2021-11-28T02:53:01.942062Z	critical	envoy lua	script log:  - content-length: 154
2021-11-28T02:53:01.942065Z	critical	envoy lua	script log:  - accept: application/json
2021-11-28T02:53:01.942067Z	critical	envoy lua	script log:  - content-type: application/json
2021-11-28T02:53:01.942069Z	critical	envoy lua	script log:  - accept-encoding: gzip
2021-11-28T02:53:01.942071Z	critical	envoy lua	script log:  - x-forwarded-proto: http
2021-11-28T02:53:01.942074Z	critical	envoy lua	script log:  - x-request-id: 1f9c937e-6532-4116-82c9-3e998becfae3
2021-11-28T02:53:01.942076Z	critical	envoy lua	script log:  - x-envoy-decorator-operation: kratos-public.identity.svc.cluster.local:80/*
2021-11-28T02:53:01.942172Z	critical	envoy lua	script log: {"csrf_token":"BHXH+FQV0FpeM+Y/fQo7IRmXt0a6vy8jnQzJRFqPrcQ=","method":"password","password":"somepassword","traits":{"email":"some@email.com"}}
```

**Comment by aeneasr** [comment]
You need to forward the client‘s (presumably the browser calling your API) HTTP Request Headers, which should include a Cookie Header! Only that way is the server able to verify if the sent CSRF token matches the expected CSRF token.

**Comment by brycesteinhoff** [comment]
I was wondering the same. It’s just strange that the Go SDK has a
SetCsrfToken method, but it doesn’t seem to respect the value set there. Do
you know if there’s a way with the Go SDK to set headers on the request?

**Comment by aeneasr** [comment]
You wouldn't actually call

```
successfulSelfServiceRegistration, httpResponse, err := client.kratosPublicClient.V0alpha2Api.
	SubmitSelfServiceRegistrationFlow(context.Background()).
	Flow(flowId).
	SubmitSelfServiceRegistrationFlowBody(submitSelfServiceRegistrationFlowBody).
	Execute()
```

from your backend unless you were performing an API-based flow. The idea is that you render an HTML form in the browser

```
<form action="https://kratos-goes-here/...">
  <input name="traits.email" ...>
  ...
</form>
```

so the method used is not intended to be used like that!

**Comment by aeneasr** [comment]
For API flows check: https://github.com/ory/kratos/tree/master/examples/go

**Comment by brycesteinhoff** [comment]
The SDK provides other browser flows explicitly, and the documentation
alludes to this type of setup, so I think it makes sense. I don’t
anticipate exposing Kratos directly outside my infrastructure. It seems
like maybe the SDK is missing this functionality.

**Comment by brycesteinhoff** [comment]
Thanks again for your responses and guidance @aeneasr.

For context, I'm building my own API around Kratos and Hydra so that I can integrate them and expose the precise interface I'd like for my app. I'm using gRPC from the browser for other aspects of the platform, and would like to use the same API machinery on my client+server sides for identity/auth stuff too.

I understand Kratos server-side needs the CSRF token in Cookie header, to ensure it came from the browser (which it can trust since it sets it HTTP-only). I want to keep this security measure working, of course, but was hoping there was some magic under the hood in the Go SDK to translate the `csrf_token` passed in via `SetCsrfToken` into the Cookie header.

Since the Go SDK would likely be used server-side and since the Kratos API accepts (requires, actually) the `csrf_token` value in the body, I assumed (hoped? :stuck_out_tongue: ) that it would facilitate this. Out of curiosity, what is the purpose of requiring `csrf_token` in the body payload, if the Kratos server requires the Cookie header to be present?

Digging into the source more, I see that there isn't that magic, or any way to manipulate the HTTP request before it's sent:

https://github.com/ory/kratos-client-go/blob/v0.8.0-alpha.2/api_v0alpha2.go#L5724

https://github.com/ory/kratos-client-go/blob/v0.8.0-alpha.2/client.go#L196

Maybe it would be good to include a way to set the Cookie header in the Go SDK (from `SetCsrfToken` method), or expose the HTTP request for manipulation before it's sent?

For my present use case, I'll figure out a different solution to make it work. Either I'll not use the SDK for this request or potentially add my own magic to translate the the body `csrf_token` into Cookie header in my service mesh layer.

**Comment by brycesteinhoff** [comment]
Looking at the Go client source more, I see it's possible to add some default headers to requests:

https://github.com/ory/kratos-client-go/blob/v0.8.0-alpha.2/configuration.go#L120

https://github.com/ory/kratos-client-go/blob/v0.8.0-alpha.2/client.go#L352

And even possible to replace the underlying `http.Client`, alternatively:

https://github.com/ory/kratos-client-go/blob/v0.8.0-alpha.2/configuration.go#L98

I'll give the default headers a shot for sending the Cookie header for these requests. I'll obviously need to instantiate unique clients for each request where I need to do this, which isn't as ideal as being able to set the header on an individual [request](https://github.com/ory/kratos-client-go/blob/v0.8.0-alpha.2/api_v0alpha2.go#L5660), but seems like it would work fine.

**Comment by aeneasr** [comment]
@brycesteinhoff sorry for dropping the ball here - is this still an issue for you?

**Comment by iakrevetkho** [comment]
Hi!

**Comment by aeneasr** [comment]
Please use the HTML form directly to submit the flow, and don’t use the Go SDK as an intermediate. For more information please see here:

- https://www.ory.sh/docs/guides/protect-page-login/go
- https://www.ory.sh/docs/kratos/concepts/ui-user-interface#ui-payloads

If you want to use the Go SDK anyways (we do not recommend this) to „proxy“ the HTML form, you’ll need to forward the cookies im the request. For more info on the request payload, check:

https://www.ory.sh/docs/kratos/self-service/flows/user-login#login-with-usernameemail-and-password

For any follow up questions please open a discussion or ask on Slack as we reserve issues for technical issues, bugs, or feature requests :)

**Comment by vizv** [comment]
I only fixed the login endpoint, the fix should work for registration as well.

Steps to fix:
1. add comment to `submitSelfServiceRegistrationFlow` of `selfservice/flow/registration/handler.go` as I did
2. regenerate API spec with `make sdk`
3. download `ory/sdk` repo and docker image, copy the spec generated to `spec/kratos` of sdk repo, and regenerate SDK for go
4. push the generated go client SDK to your fork
5. replace repo with your fork in `go.mod`, and update your code to forward CSRF-related Cookie entry
